### PR TITLE
feat: training pipeline, GPU backends, dataset recording, and observability

### PR DESCRIPTION
**Pull request by cagataycali**
## Summary

Complete training infrastructure: **6 trainer implementations**, **dataset recording** in LeRobot format, **Isaac Sim/Lab** and **Newton** GPU backends, **Robometer** benchmarking, **telemetry streaming**, and supporting utilities.

### Training Pipeline (`training/__init__.py` — 1800 lines)
| Trainer | Description |
|---------|-------------|
| `Gr00tTrainer` | NVIDIA GR00T N1.5/N1.6 fine-tuning |
| `LerobotTrainer` | HuggingFace LeRobot (ACT, Diffusion, Pi0, SmolVLA) |
| `DreamgenIdmTrainer` | GR00T-Dreams inverse dynamics model |
| `DreamgenVlaTrainer` | GR00T-Dreams VLA fine-tuning |
| `CosmosTrainer` | NVIDIA Cosmos world model |
| `CosmosTransferTrainer` | Sim-to-real domain transfer |

### Dataset Recording
- LeRobot v2/v3 format with multi-camera synchronized capture
- Episode management, auto-indexing, HuggingFace Hub upload

### GPU Simulation Backends
**Isaac Sim/Lab**: IsaacSimBackend, IsaacGymEnv (4096+ parallel envs), IsaacLabTrainer, USD↔MuJoCo bridge
**Newton**: Differentiable physics, gradient-based optimization, NewtonGymEnv

### Observability
- **Robometer**: Policy benchmarking with reward models
- **Telemetry**: Real-time streaming (WebSocket/Zenoh/OpenTelemetry)
- **Video**: Encoding (PyAV/imageio), **Visualizer**: Trajectory replay
- **Dashboard**: WebSocket-based monitoring UI
- **Processor**: HF processor → policy observation bridge
- **Kinematics**, **Motion Library**, **Stereo depth**, **Cosmos Transfer**

### Tests — 26 passed

### CODE REVIEW

**Review by awsarron — COMMENTED** [review]
## 🤖 Agent-Assisted Review — @awsarron

Thanks for this massive contribution, @cagataycali! This is an impressive amount of infrastructure — training pipeline, GPU backends, dataset recording, telemetry, and more. I've reviewed the full diff (23.5k lines across 48 files) and run the test suite locally (26/26 passing). Here's my feedback.

---

### Overall Architecture

The modular design is solid — each subsystem (training, Newton, Isaac, telemetry, Marble, etc.) is self-contained with lazy imports so the base package stays lightweight. The `Robot()` factory cleanly dispatches across MuJoCo/Isaac/Newton backends, and the `create_trainer()` factory provides a nice unified entry point for the training matrix. Good use of the Strategy pattern throughout (telemetry tiers, transport backends, solver selection).

The `pyproject.toml` restructuring to move heavy deps (`torch`, `lerobot`) into optional dependency groups is the right call — keeps `pip install strands-robots` fast for users who only need the core.

---

### Items for Consideration

#### 1. Test depth — import-only tests for 23k lines of code

All 26 tests are import/existence checks (`assert X is not None`, `assert callable(X)`). There's no behavioral testing — no unit tests for `create_trainer()` argument forwarding, `TrainConfig` field extraction, `DatasetRecorder.add_frame()` logic, `TelemetryStream.emit()` buffering, `_resolve_name()` alias resolution, etc.

I understand that many of these modules depend on heavy GPU runtimes (Isaac Sim, Newton/Warp, CUDA), but there's a lot of pure-Python logic that could be tested without any GPU deps:
- `create_trainer()` config extraction and provider dispatch
- `_resolve_name()` / `_ALIASES` in `factory.py`
- `DatasetRecorder._build_features()` feature dict construction
- `TelemetryStream` emit/flush lifecycle (with mock transports)
- `resolve_policy()` in `policy_resolver.py`
- `_infer_config_name()` in the Cosmos trainers

Not blocking on this, but would be great to add in a follow-up.

#### 2. Large single-file modules

Several files are quite large:
- `newton_backend.py` — 1,964 lines
- `marble/__init__.py` — 1,953 lines
- `training/__init__.py` — 1,890 lines
- `asset_converter.py` — 1,602 lines
- `cosmos_transfer/__init__.py` — 1,405 lines

These are manageable for now, but as the codebase grows, splitting some of these would help with maintainability. For example, `training/__init__.py` could have each trainer in its own file with the factory in `__init__.py`. Non-blocking — just noting for future consideration.

#### 3. Broad exception handling in several modules

There are quite a few bare `except Exception:` blocks (especially in `cosmos_transfer/`, `marble/`, and `dataset_recorder.py`) that silently swallow errors. Some of these log warnings, which is good, but others just pass silently. For a robotics framework where debugging failures matters, it would be worth tightening these up over time — catching specific exceptions where possible.

#### 4. `subprocess.run` without timeout

The training pipeline launches external processes (`gr00t.experiment.launch_finetune`, `lerobot.scripts.lerobot_train`, Cosmos scripts, etc.) via `subprocess.run()` without a `timeout` parameter. For long-running training jobs this is probably fine (you want them to run to completion), but for the evaluation subprocess calls it might be worth adding a timeout to prevent hangs.

#### 5. `evaluate()` function — event loop lifecycle

In `training/__init__.py`, the `evaluate()` function creates a new `asyncio.event_loop` for async policies:

```python
_async_loop = None
if inspect.iscoroutinefunction(getattr(policy, "get_actions", None)):
    _async_loop = asyncio.new_event_loop()
```

This works but could conflict if called from within an existing async context. Minor concern — just noting it.

#### 6. Newton procedural robots — limited set

The `_PROCEDURAL_ROBOTS` dict in `newton_backend.py` only covers `so100`, `so101`, and `koch`. The factory's `_UNIFIED_ROBOTS` has 35+ robots. When a user does `Robot("unitree_g1", backend="newton")` and no URDF is found, the fallback behavior should be clearly documented or error out gracefully rather than silently creating a generic chain.

#### 7. `training/__init__.py` unused imports

Lines 43-44 have `field` from dataclasses and `List` from typing imported with `# noqa: F401` but they don't appear to be used in this file. Minor cleanup.

---

### Things I Like

- **Lazy imports everywhere** — the module loads fast even with 48 new files. Heavy deps (torch, warp, isaac, robometer, scipy) are only imported at point of use.
- **`_discover_nvidia_cuda_lib_paths()`** — thoughtful handling of the CUDA library path mess when pip-installed NVIDIA packages conflict with system CUDA. This will save users real debugging time.
- **Telemetry design** — using `collections.deque` instead of `asyncio.Queue` for the 50Hz control loop is the right call. The tier-based routing (STREAM/BATCH/STORAGE) with per-tier batch configs is well thought out.
- **`DreamgenIdmTrainer` unknown param warning** — nice touch warning users about commonly misused parameters from Issue #73 rather than silently dropping them.
- **`CosmosTrainer` / `CosmosTransferTrainer` warnings** — the `UserWarning` for unverified LoRA/backbone params is honest and helpful. Better than pretending they work.
- **`DatasetRecorder`** — clean bridge between the control loop and LeRobot v3 format. The `inspect.signature` check for `streaming_encoding` compatibility across LeRobot versions is a nice defensive pattern.
- **pyproject.toml** — moving `torch` and `lerobot` out of core deps into optional groups is a significant improvement for install ergonomics.

---

### Summary

This is a substantial and well-structured addition to the robotics infrastructure. The architecture decisions are sound — lazy imports, factory patterns, clean abstractions. The main gap is test coverage beyond import checks, which I'd encourage addressing in a follow-up PR. The items above are all non-blocking suggestions.

Nice work! 🚀

---

*AI agent review on behalf of @awsarron. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

**Review by awsarron — DISMISSED** [review]
## Follow-up Review — @awsarron

Thanks for the quick turnaround on the fixes, @cagataycali. I pulled the latest (`c17d0fc`) and re-reviewed the full diff.

### Resolved items

| # | Item | Status |
|---|------|--------|
| 4 | `subprocess.run` without timeout | ✅ `timeout=3600` added to `_evaluate_in_process()` — confirmed at line 496 |
| 7 | Unused imports (`field`, `List`) | ✅ Removed cleanly |

Both fixes verified locally.

### Verification

- `pytest tests/test_training_pipeline.py` — 26/26 passing
- `create_trainer("groot", max_steps=500, output_dir="./test", batch_size=8)` — config forwarding works correctly (the `create_trainer()` fix from commit `2b69897` properly extracts `TrainConfig` fields from kwargs)
- `create_trainer("lerobot", config=TrainConfig(max_steps=100), max_steps=999)` — kwargs override explicit config fields as expected
- `resolve_policy()` — tested mock, HF model IDs, server addresses, WebSocket URLs — all resolve correctly
- `_resolve_name()` — alias resolution works (g1→unitree_g1, franka→panda, SO100→so100)
- No dangerous patterns found (`eval()`, `exec()`, `pickle`, `shell=True`)
- Lint is clean

### A few more things I noticed on this pass

1. **Dead imports in `dreamgen/__init__.py`** — `Path` (line 47), `subprocess` (line 142), and `sys` (line 143) are imported with `noqa: F401` but none are actually used in the module. Minor cleanup.

2. **`Gr00tTrainer.train()` and other training subprocess calls still have no timeout** — Lines 249, 454, 596, 702, 1080, 1504 all call `subprocess.run()` without `timeout`. I understand the reasoning (training jobs should run to completion), and the eval path now has a timeout, so this is fine for now. Just noting it for awareness — if any of these are ever called from an automated pipeline, a generous timeout (e.g. 24h) would be a safety net.

3. **`evaluate()` event loop** — The `asyncio.new_event_loop()` pattern at line 1560 works but will raise `RuntimeError` if called from within an existing async context (e.g. a Jupyter notebook or an async agent loop). A defensive check like `asyncio.get_running_loop()` wrapped in a try/except could handle that case. Low priority — just noting it.

4. **Broad `except Exception` in `cosmos_transfer/` and `marble/`** — These modules use fallback chains (try method A → try method B → fallback C) where broad catches make sense for the pattern. The ones that log with `exc_info=True` are fine. The bare `except Exception: pass` in `cosmos_transfer/__init__.py` line 1306 (the `__del__` cleanup) is also fine — swallowing errors in destructors is standard practice.

### Summary

The fixes from `3144c9a` address the items I flagged. The `create_trainer()` config forwarding fix from `2b69897` is a nice improvement — users can now pass `max_steps`, `output_dir`, etc. directly without constructing a `TrainConfig` manually. The acknowledged items (test depth, module splitting, exception tightening, Newton fallback) are reasonable follow-ups.

Approving — nice work on the quick iteration. 🚀

---

*AI agent review on behalf of @awsarron. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*
